src/transformations/pivot.ts

```diff
@@ -0,0 +1,214 @@
+ * @param valuesTo name of attribute under which *values* of the toPivot
+ *  attributes will go
+ * @returns pivoted dataset
+ */
+export function pivotLonger(
+  dataset: DataSet,
+  toPivot: string[],
+  namesTo: string,
+  valuesTo: string
+): DataSet {
+  // TODO: is this a necessary requirement?
+  if (dataset.collections.length !== 1) {
+    throw new Error(
+      `pivot longer can only be used on a single-collection dataset`
+    );
+  }
+
+  // remove pivoting attributes
+  const collection = { ...dataset.collections[0] };
+  collection.attrs =
+    collection.attrs?.filter((attr) => !toPivot.includes(attr.name)) || [];
+
+  // NOTE: do not copy formulas: dependencies may be removed by the pivot
+  eraseFormulas(collection.attrs);
+
+  // add namesTo and valuesTo attributes
+  // NOTE: valuesTo might hold values of different types
+  // so we can't be sure it's either numeric / categorical
+  collection.attrs.push(
+    {
+      name: namesTo,
+      type: "categorical",
+    },
+    {
+      name: valuesTo,
+    }
+  );
+
+  const records = [];
+  for (const toPivotAttr of toPivot) {
+    for (const record of dataset.records) {
+      // remove attributes being pivoted
+      const shortRecord = removeFields({ ...record }, toPivot);
+
+      // put pivoted attribute name under namesTo attribute,
+      // value of record at toPivotAttr under valuesTo attribute
+      shortRecord[namesTo] = toPivotAttr;
+      shortRecord[valuesTo] = record[toPivotAttr];
+
+      records.push(shortRecord);
+    }
+  }
+
+  return {
+    collections: [collection],
+    records,
+  };
+}
+
+/**
+ * Extracts the values of the namesFrom attribute into new attributes,
+ * with the values from the valuesFrom attribute as their values. The dataset
+ * gets "wider" (more attributes), but less long (fewer cases).
+ *
+ * @param dataset the dataset to pivot
+ * @param namesFrom name of attribute from which to extract new attribute names
+ * @param valuesFrom name of attribute holding the values that will go
+ *  under the new attribute names
+ */
+export function pivotWider(
+  dataset: DataSet,
+  namesFrom: string,
+  valuesFrom: string
+): DataSet {
+  // TODO: is this a necessary requirement?
+  if (dataset.collections.length !== 1) {
+    throw new Error(
+      `pivot wider can only be used on a single-collection dataset`
+    );
+  }
+
+  const collection = { ...dataset.collections[0] };
+
+  // get list of names to make attributes for
+  const newAttrs = Array.from(
+    new Set(
+      dataset.records.map((rec) => {
+        if (rec[namesFrom] === undefined) {
+          throw new Error(
+            `invalid attribute to retrieve names from: ${namesFrom}`
+          );
+        }
+        if (typeof rec[namesFrom] === "object") {
+          throw new Error(
+            `cannot use object values (${namesFrom}) as attribute names`
+          );
+        }
+
+        return rec[namesFrom] as string;
```

Comment:
Can we instead check

```typescript
if (!(typeof rec[namesFrom] === "string")) {
  throw new Error("Expected string, instead got ...");
}
```
Not a big deal, but want to avoid casting.

Reply:
I think we do want to allow things like numbers becoming attribute names though, right? E.g. if the numbers are really categorical in a sense. So maybe we should use `String(rec[namesFrom])` instead? @stardust66

Reply:
Oh that's right. Yeah, using `String(rec[namesFrom])` makes sense.